File: GetDown/standalone/random_movement.py
import asyncio
import json
import random
import math
import sys
import websockets

API_URL = "ws://localhost:8004"
PLUGIN_NAME = "RandomMovementGenerator"
PLUGIN_DEVELOPER = "User"
DURATION = 5.0
FPS = 30
FRAME_INTERVAL = 1.0 / FPS

# Parameters to randomize and their ranges
PARAMS = [
    # (parameter_id, min_val, max_val, speed_factor)
    # speed_factor controls how fast the parameter changes (lower = smoother)
    # Cranked WAY beyond normal ranges for maximum distortion
    ("FaceAngleX", -180, 180, 10.0),
    ("FaceAngleY", -180, 180, 10.0),
    ("FaceAngleZ", -180, 180, 10.0),
    ("FacePositionX", -50, 50, 10.0),
    ("FacePositionY", -50, 50, 10.0),
    ("EyeOpenLeft", -5, 5, 15.0),
    ("EyeOpenRight", -5, 5, 15.0),
    ("EyeLeftX", -10, 10, 15.0),
    ("EyeLeftY", -10, 10, 15.0),
    ("EyeRightX", -10, 10, 15.0),
    ("EyeRightY", -10, 10, 15.0),
    ("MouthOpen", -3, 5, 15.0),
    ("MouthSmile", -10, 10, 10.0),
    ("MouthForm", -10, 10, 10.0),
    ("BrowLeftY", -10, 10, 10.0),
    ("BrowRightY", -10, 10, 10.0),
    ("BodyAngleX", -90, 90, 8.0),
    ("BodyAngleY", -90, 90, 8.0),
    ("BodyAngleZ", -90, 90, 8.0),
    # Extra params that many models support
    ("CheekPuff", -5, 5, 10.0),
    ("TongueOut", -3, 5, 10.0),
    ("EyeSquintLeft", -5, 5, 10.0),
    ("EyeSquintRight", -5, 5, 10.0),
]

# How often to pick new random targets (seconds) — lower = more erratic
TARGET_CHANGE_INTERVAL = 0.03
# No smoothing — jump straight to random values for maximum horror
NO_LERP = True


async def send_request(ws, msg_type, data=None):
    request = {
        "apiName": "VTubeStudioPublicAPI",
        "apiVersion": "1.0",
        "requestID": str(random.randint(1000, 9999)),
        "messageType": msg_type,
    }
    if data:
        request["data"] = data
    await ws.send(json.dumps(request))
    response = json.loads(await ws.recv())
    return response


async def authenticate(ws):
    # Request auth token
    print("Requesting authentication token...")
    resp = await send_request(ws, "AuthenticationTokenRequest", {
        "pluginName": PLUGIN_NAME,
        "pluginDeveloper": PLUGIN_DEVELOPER,
    })

    if resp.get("messageType") == "APIError":
        print(f"Error: {resp['data']['message']}")
        sys.exit(1)

    token = resp["data"]["authenticationToken"]
    print("Got token. Please APPROVE the plugin in VTube Studio if prompted...")
    await asyncio.sleep(2)

    # Authenticate with token
    resp = await send_request(ws, "AuthenticationRequest", {
        "pluginName": PLUGIN_NAME,
        "pluginDeveloper": PLUGIN_DEVELOPER,
        "authenticationToken": token,
    })

    if not resp["data"].get("authenticated"):
        print("Authentication failed. Did you approve the plugin in VTube Studio?")
        sys.exit(1)

    print("Authenticated!")


def lerp(a, b, t):
    return a + (b - a) * t


async def get_model_parameters(ws):
    """Query the model's actual input parameters and their ranges."""
    print("Querying model parameters...")
    resp = await send_request(ws, "InputParameterListRequest")
    if resp.get("messageType") == "APIError":
        print(f"Warning: Could not get parameters: {resp['data']['message']}")
        return None
    data = resp["data"]
    # Find the right key for parameters
    params = data.get("modelParameters") or data.get("defaultParameters") or data.get("customParameters")
    if not params:
        print(f"Response keys: {list(data.keys())}")
        # Try combining all parameter lists we can find
        all_params = []
        for key, val in data.items():
            if isinstance(val, list):
                all_params.extend(val)
        if all_params:
            params = all_params
        else:
            print("Could not find parameters in response. Falling back to hardcoded list.")
            return None
    print(f"Model has {len(params)} input parameters:")
    for p in params:
        print(f"  {p['name']}: {p['min']} to {p['max']} (default {p['defaultValue']})")
    return params


async def run_random_movements(ws):
    model_params = await get_model_parameters(ws)
    if not model_params:
        print("Could not get model parameters. Exiting.")
        sys.exit(1)

    # Build a dict of param name -> {min, max}
    param_info = {}
    for p in model_params:
        param_info[p["name"]] = {"min": p["min"], "max": p["max"]}

    print(f"\nUsing {len(param_info)} parameters from model")
    print(f"Starting UNNATURAL movements for {DURATION} seconds at {FPS} FPS...")
    print("Make sure OBS is recording!\n")

    # Identify paired parameters for desync (left vs right)
    left_right_pairs = []
    seen = set()
    for name in param_info:
        nl = name.lower()
        if "left" in nl:
            right_name = name.replace("Left", "Right").replace("left", "right")
            if right_name in param_info and name not in seen:
                left_right_pairs.append((name, right_name))
                seen.add(name)
                seen.add(right_name)

    print(f"Found {len(left_right_pairs)} left/right pairs to desync:")
    for l, r in left_right_pairs:
        print(f"  {l} <-> {r}")

    elapsed = 0.0
    frame = 0
    total_frames = int(DURATION * FPS)

    while elapsed < DURATION:
        frame_start = asyncio.get_event_loop().time()
        t = elapsed / DURATION  # 0..1 progress
        frame += 1

        param_values = []

        for name, info in param_info.items():
            lo, hi = info["min"], info["max"]
            mid = (lo + hi) / 2.0
            rng = hi - lo
            if rng == 0:
                rng = 1

            # Different chaos techniques applied simultaneously
            technique = random.choice(["snap", "oscillate", "sine_stack", "hold_extreme"])

            if technique == "snap":
                # Snap between min and max with no in-between
                val = lo if random.random() < 0.5 else hi
            elif technique == "oscillate":
                # High-frequency oscillation at different rates per param
                freq = random.uniform(8, 40)
                val = mid + (rng / 2) * math.sin(elapsed * freq * 2 * math.pi)
            elif technique == "sine_stack":
                # Stack multiple sine waves for organic-but-wrong movement
                val = mid
                for i in range(4):
                    freq = random.uniform(3, 25)
                    val += (rng / 4) * math.sin(elapsed * freq + i * 1.7)
                val = max(lo, min(hi, val))
            else:
                # Hold at extreme value
                val = lo if frame % 7 < 3 else hi

            param_values.append({"id": name, "value": val})

        # DESYNC: Force left/right pairs to opposite extremes
        for left_name, right_name in left_right_pairs:
            lo_l, hi_l = param_info[left_name]["min"], param_info[left_name]["max"]
            lo_r, hi_r = param_info[right_name]["min"], param_info[right_name]["max"]

            # Alternate which side is at which extreme, at different rates
            phase = math.sin(elapsed * random.uniform(5, 15))
            if phase > 0:
                left_val, right_val = hi_l, lo_r
            else:
                left_val, right_val = lo_l, hi_r

            # Sometimes make them both slam to the same extreme (also uncanny)
            if random.random() < 0.2:
                left_val = right_val = hi_l

            # Overwrite the values for these params
            for pv in param_values:
                if pv["id"] == left_name:
                    pv["value"] = left_val
                elif pv["id"] == right_name:
                    pv["value"] = right_val

        # Send WITHOUT waiting for response (fire-and-forget for speed)
        request = {
            "apiName": "VTubeStudioPublicAPI",
            "apiVersion": "1.0",
            "requestID": str(frame),
            "messageType": "InjectParameterDataRequest",
            "data": {
                "faceFound": True,
                "mode": "set",
                "parameterValues": param_values,
            },
        }
        await ws.send(json.dumps(request))
        # Drain response without blocking
        try:
            await asyncio.wait_for(ws.recv(), timeout=0.005)
        except asyncio.TimeoutError:
            pass

        elapsed += FRAME_INTERVAL
        progress = int((elapsed / DURATION) * 50)
        bar = "=" * progress + " " * (50 - progress)
        print(f"\r  [{bar}] {elapsed:.1f}s / {DURATION:.1f}s", end="", flush=True)

        frame_time = asyncio.get_event_loop().time() - frame_start
        sleep_time = max(0, FRAME_INTERVAL - frame_time)
        await asyncio.sleep(sleep_time)

    print("\nDone!")


async def main():
    print(f"Connecting to VTube Studio at {API_URL}...")
    try:
        async with websockets.connect(API_URL) as ws:
            await authenticate(ws)
            await asyncio.sleep(1)
            await run_random_movements(ws)
    except ConnectionRefusedError:
        print("Could not connect to VTube Studio.")
        print("Make sure VTube Studio is running and the API is enabled:")
        print("  Settings > General Settings > Start API (Port 8004)")
        sys.exit(1)


if __name__ == "__main__":
    asyncio.run(main())

File: GetDown/standalone/src/random_movement.ts
import "dotenv/config";
import WebSocket from "ws";
import {
  TwitchChatManager,
  ChatCommandManager,
} from "../../../utils/chatgod-js/src/services/TwitchChatManager.js";

const API_URL = "ws://localhost:8004";
const PLUGIN_NAME = "RandomMovementGenerator";
const PLUGIN_DEVELOPER = "Sarxina";
const TRIGGER_COMMAND = "!getdown";
const FPS = 30;
const FRAME_INTERVAL = 1000 / FPS; // ms

let requestId = 0;

function nextId(): string {
  return String(++requestId);
}

// --- Terminal UI helpers ---

function clearScreen(): void {
  process.stdout.write(process.platform === "win32" ? "\x1Bc" : "\x1B[2J\x1B[H");
}

function printHeader(): void {
  console.log("\x1B[1;36m" + "=".repeat(50));
  console.log("  GetDown - Random Movement Generator");
  console.log("=".repeat(50) + "\x1B[0m");
  console.log();
  console.log("  Press \x1B[1;33mCtrl+C\x1B[0m to stop");
  console.log();
}

interface ParamValue {
  id: string;
  value: number;
  min: number;
  max: number;
}

function printParams(paramValues: ParamValue[], frame: number): void {
  // Move cursor to line 7 (after header)
  process.stdout.write("\x1B[7;0H");
  process.stdout.write(`  \x1B[1mFrame ${frame}\x1B[0m\n`);
  process.stdout.write(`  ${"─".repeat(46)}\n`);

  for (const pv of paramValues) {
    const { id: name, value: val, min, max } = pv;

    // Color based on value magnitude
    let color: string;
    if (Math.abs(val) > 50) {
      color = "\x1B[1;31m"; // red
    } else if (Math.abs(val) > 10) {
      color = "\x1B[1;33m"; // yellow
    } else {
      color = "\x1B[0;37m"; // white
    }

    // Simple bar visualization
    const barWidth = 20;
    const barCenter = Math.floor(barWidth / 2);
    const norm = Math.max(-1, Math.min(1, val / Math.max(Math.abs(max || 100), 1)));
    let barPos = Math.floor(barCenter + norm * barCenter);
    barPos = Math.max(0, Math.min(barWidth - 1, barPos));

    const bar = Array(barWidth).fill("·");
    bar[barCenter] = "│";
    bar[barPos] = "█";
    const barStr = bar.join("");

    process.stdout.write(
      `  ${color}${name.padEnd(22)}\x1B[0m [${barStr}] ${color}${val.toFixed(2).padStart(8)}\x1B[0m\n`
    );
  }

  // Clear remaining lines
  process.stdout.write("\x1B[J");
}

// --- VTube Studio API ---

async function sendRequest(ws: WebSocket, messageType: string, data?: Record<string, unknown>): Promise<any> {
  const request: Record<string, unknown> = {
    apiName: "VTubeStudioPublicAPI",
    apiVersion: "1.0",
    requestID: nextId(),
    messageType,
  };
  if (data) request.data = data;

  return new Promise((resolve, reject) => {
    const handler = (raw: WebSocket.RawData) => {
      const resp = JSON.parse(raw.toString());
      if (resp.requestID === request.requestID) {
        ws.off("message", handler);
        resolve(resp);
      }
    };
    ws.on("message", handler);
    ws.send(JSON.stringify(request));
    setTimeout(() => {
      ws.off("message", handler);
      reject(new Error(`Request ${messageType} timed out`));
    }, 10000);
  });
}

async function authenticate(ws: WebSocket): Promise<void> {
  console.log("  Requesting authentication token...");
  const tokenResp = await sendRequest(ws, "AuthenticationTokenRequest", {
    pluginName: PLUGIN_NAME,
    pluginDeveloper: PLUGIN_DEVELOPER,
  });

  if (tokenResp.messageType === "APIError") {
    console.error(`  Error: ${tokenResp.data.message}`);
    process.exit(1);
  }

  const token = tokenResp.data.authenticationToken;
  console.log("  Approve the plugin in VTube Studio if prompted...");
  await new Promise((r) => setTimeout(r, 2000));

  const authResp = await sendRequest(ws, "AuthenticationRequest", {
    pluginName: PLUGIN_NAME,
    pluginDeveloper: PLUGIN_DEVELOPER,
    authenticationToken: token,
  });

  if (!authResp.data?.authenticated) {
    console.error("  Authentication failed. Did you approve the plugin in VTube Studio?");
    process.exit(1);
  }

  console.log("  Authenticated!");
}

interface ModelParam {
  name: string;
  min: number;
  max: number;
}

async function getModelParameters(ws: WebSocket): Promise<ModelParam[] | null> {
  const resp = await sendRequest(ws, "InputParameterListRequest");
  if (resp.messageType === "APIError") return null;

  const data = resp.data;
  let params: any[] | null =
    data.modelParameters ?? data.defaultParameters ?? data.customParameters ?? null;

  if (!params) {
    const allParams: any[] = [];
    for (const val of Object.values(data)) {
      if (Array.isArray(val)) allParams.push(...val);
    }
    params = allParams.length > 0 ? allParams : null;
  }

  return params;
}

// --- Movement logic ---

async function runRandomMovements(ws: WebSocket): Promise<void> {
  const modelParams = await getModelParameters(ws);
  if (!modelParams) {
    console.error("  Could not get model parameters. Exiting.");
    process.exit(1);
  }

  const paramInfo = new Map<string, { min: number; max: number }>();
  for (const p of modelParams) {
    paramInfo.set(p.name, { min: p.min, max: p.max });
  }

  // Find left/right pairs for desync
  const leftRightPairs: [string, string][] = [];
  const seen = new Set<string>();
  for (const name of paramInfo.keys()) {
    const nl = name.toLowerCase();
    if (nl.includes("left")) {
      const rightName = name.replace("Left", "Right").replace("left", "right");
      if (paramInfo.has(rightName) && !seen.has(name)) {
        leftRightPairs.push([name, rightName]);
        seen.add(name);
        seen.add(rightName);
      }
    }
  }

  clearScreen();
  printHeader();

  let frame = 0;
  let elapsed = 0;
  let timerId: ReturnType<typeof setTimeout> | null = null;

  stopFn = () => {
    if (timerId) clearTimeout(timerId);
  };

  const techniques = ["snap", "oscillate", "sine_stack", "hold_extreme"] as const;

  const tick = (): void => {
    if (!running) return;
    const frameStart = performance.now();
    frame++;

    const paramValues: ParamValue[] = [];

    for (const [name, info] of paramInfo) {
      const { min: lo, max: hi } = info;
      const mid = (lo + hi) / 2;
      let rng = hi - lo;
      if (rng === 0) rng = 1;

      const technique = techniques[Math.floor(Math.random() * techniques.length)];
      let val: number;

      switch (technique) {
        case "snap":
          val = Math.random() < 0.5 ? lo : hi;
          break;
        case "oscillate": {
          const freq = Math.random() * 32 + 8;
          val = mid + (rng / 2) * Math.sin(elapsed * freq * 2 * Math.PI);
          break;
        }
        case "sine_stack": {
          val = mid;
          for (let i = 0; i < 4; i++) {
            const freq = Math.random() * 22 + 3;
            val += (rng / 4) * Math.sin(elapsed * freq + i * 1.7);
          }
          val = Math.max(lo, Math.min(hi, val));
          break;
        }
        case "hold_extreme":
          val = frame % 7 < 3 ? lo : hi;
          break;
      }

      paramValues.push({ id: name, value: val, min: lo, max: hi });
    }

    // Desync left/right pairs
    for (const [leftName, rightName] of leftRightPairs) {
      const loL = paramInfo.get(leftName)!.min;
      const hiL = paramInfo.get(leftName)!.max;
      const loR = paramInfo.get(rightName)!.min;
      const hiR = paramInfo.get(rightName)!.max;

      const phase = Math.sin(elapsed * (Math.random() * 10 + 5));
      let leftVal: number, rightVal: number;
      if (phase > 0) {
        leftVal = hiL;
        rightVal = loR;
      } else {
        leftVal = loL;
        rightVal = hiR;
      }

      if (Math.random() < 0.2) {
        leftVal = rightVal = hiL;
      }

      for (const pv of paramValues) {
        if (pv.id === leftName) pv.value = leftVal;
        else if (pv.id === rightName) pv.value = rightVal;
      }
    }

    // Send to VTube Studio
    const request = {
      apiName: "VTubeStudioPublicAPI",
      apiVersion: "1.0",
      requestID: String(frame),
      messageType: "InjectParameterDataRequest",
      data: {
        faceFound: true,
        mode: "set",
        parameterValues: paramValues.map((pv) => ({ id: pv.id, value: pv.value })),
      },
    };
    ws.send(JSON.stringify(request));

    // Update terminal display
    printParams(paramValues, frame);

    elapsed += FRAME_INTERVAL / 1000;
    const frameTime = performance.now() - frameStart;
    const sleepTime = Math.max(0, FRAME_INTERVAL - frameTime);
    timerId = setTimeout(tick, sleepTime);
  };

  tick();
}

let running = false;
let stopFn: (() => void) | null = null;

async function main(): Promise<void> {
  console.log(`  Connecting to VTube Studio at ${API_URL}...`);

  const ws = new WebSocket(API_URL);

  await new Promise<void>((resolve, reject) => {
    ws.on("open", resolve);
    ws.on("error", () => {
      console.error("  Could not connect to VTube Studio.");
      console.error("  Make sure VTube Studio is running and the API is enabled.");
      console.error(`  Settings > General Settings > Start API (check port matches ${API_URL})`);
      process.exit(1);
    });
  });

  await authenticate(ws);

  const chatManager = new TwitchChatManager(() => {});
  new ChatCommandManager(
    TRIGGER_COMMAND,
    (subcommand, chatter) => {
      const arg = subcommand.trim().toLowerCase();
      if (arg === "" || arg === "on") {
        if (!running) {
          console.log(`\n  ${chatter} started random movement!`);
          running = true;
          runRandomMovements(ws);
        }
      } else if (arg === "off") {
        if (running && stopFn) {
          console.log(`\n  ${chatter} stopped random movement.`);
          running = false;
          stopFn();
          stopFn = null;
        }
      }
    },
    chatManager
  );

  console.log(`  Listening for "${TRIGGER_COMMAND}" in Twitch chat. Ctrl+C to quit.\n`);
}

main().catch((err) => {
  console.error(err);
  process.exit(1);
});

process.on("SIGINT", () => {
  clearScreen();
  console.log("\n  Stopped.\n");
  process.exit(0);
});
